exp/runtime/api/v1beta1/extensionconfig_types.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+// ANCHOR: ExtensionConfigSpec
+
+// ExtensionConfigSpec defines the desired state of ExtensionConfig.
+type ExtensionConfigSpec struct {
+	// ClientConfig defines how to communicate with ExtensionHandlers.
+	ClientConfig ClientConfig `json:"clientConfig"`
+
+	// NamespaceSelector decides whether to run the webhook on an object based
+	// on whether the namespace for that object matches the selector.
+	// Default to the empty LabelSelector, which matches everything.
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+}
+
+// ClientConfig contains the information to make a client
+// connection with an ExtensionHandler.
+type ClientConfig struct {
+	// URL gives the location of the ExtensionHandler, in standard URL form
+	// (`scheme://host:port/path`). Exactly one of `url` or `service`
```

Comment:
This looks like a [discriminated union](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1027-api-unions/README.md). It should have a discriminator field to allow the consumer to determine which type they should be expecting.

There are some api gen tags you can add for this mentioned in the KEP

I'm not sure whether this works with OpenAPI just yet though, I think it's a feature gap in the CRD gen tool

If we do this, it's best not to include other fields (CABundle) as it can be confusing to an end user, typically you'd have the union inside another key on its own

```
clientConfig:
  endpoint:
     type: URL
     url: https://...
  caBundle: ...
```